cmd/goal/abi/abi.go

```diff
@@ -0,0 +1,1193 @@
+		}
+	}
+	if len(dynamicSegments) > 0 {
+		dynamicSegments[len(dynamicSegments)-1].right = len(valueBytes) - 1
+		iterIndex = len(valueBytes)
+	}
+	if iterIndex < len(valueBytes) {
+		return Value{}, fmt.Errorf("input byte not fully consumed")
+	}
+
+	// check segment indices are valid
+	segIndexArr := make([]int, len(dynamicSegments)*2)
+	for index, segment := range dynamicSegments {
+		segIndexArr[index*2] = segment.left
+		segIndexArr[index*2+1] = segment.right
+	}
+	for i := 0; i < len(segIndexArr); i++ {
+		if i%2 == 1 {
+			if i != len(segIndexArr)-1 && segIndexArr[i]+1 != segIndexArr[i+1] {
+				return Value{}, fmt.Errorf("dynamic segment should sit next to each other")
+			}
+		} else {
+			if segIndexArr[i] >= segIndexArr[i+1] {
+				return Value{}, fmt.Errorf("dynamic segment should display a [l, r] space")
+			}
+		}
+	}
+
+	segIndex := 0
+	for i := 0; i < len(valueType.childTypes); i++ {
+		if valuePartition[i] == nil {
+			valuePartition[i] = valueBytes[dynamicSegments[segIndex].left : dynamicSegments[segIndex].right+1]
+			segIndex++
+		}
+	}
+
+	values := make([]Value, 0)
+	for i := 0; i < len(valueType.childTypes); i++ {
+		valueTi, err := Decode(valuePartition[i], valueType.childTypes[i])
+		if err != nil {
+			return Value{}, err
+		}
+		values = append(values, valueTi)
+	}
+	return Value{
+		valueType: valueType,
+		value:     values,
+	}, nil
+}
+
+// MakeUint8 takes a go `uint8` and gives an ABI Value of ABI type `uint8`.
+func MakeUint8(value uint8) (Value, error) {
+	bigInt := big.NewInt(int64(value))
+	return MakeUint(bigInt, 8)
+}
+
+// MakeUint16 takes a go `uint16` and gives an ABI Value of ABI type `uint16`.
+func MakeUint16(value uint16) (Value, error) {
+	bigInt := big.NewInt(int64(value))
+	return MakeUint(bigInt, 16)
+}
+
+// MakeUint32 takes a go `uint32` and gives an ABI Value of ABI type `uint32`.
+func MakeUint32(value uint32) (Value, error) {
+	bigInt := big.NewInt(int64(value))
+	return MakeUint(bigInt, 32)
+}
+
+// MakeUint64 takes a go `uint64` and gives an ABI Value of ABI type `uint64`.
+func MakeUint64(value uint64) (Value, error) {
+	bigInt := big.NewInt(int64(0)).SetUint64(value)
+	return MakeUint(bigInt, 64)
+}
+
+// MakeUint takes a big integer representation and a type size,
+// and returns an ABI Value of ABI Uint<size> type.
+func MakeUint(value *big.Int, size uint16) (Value, error) {
+	typeUint, err := MakeUintType(size)
+	if err != nil {
+		return Value{}, err
+	}
+	upperLimit := big.NewInt(0).Lsh(big.NewInt(1), uint(size))
+	if value.Cmp(upperLimit) >= 0 {
+		return Value{}, fmt.Errorf("passed value larger than uint size %d", size)
+	}
+	return Value{
+		valueType: typeUint,
+		value:     value,
+	}, nil
+}
+
+// MakeUfixed takes a big rational number representation, a type size, and a type precision,
+// and returns an ABI Value of ABI UFixed<size>x<precision>
+func MakeUfixed(value *big.Rat, size uint16, precision uint16) (Value, error) {
```

Comment:
It might be an overkill to do representation of `ufixed` with `big.Rat` in this case. Although it is straightforward to consider `ufixed` with rational representation like `p / q` (where `p, q \in Nat`), but this incurs unnecessary computational cost.

It is already switched to `big.Int` to store only the numerator of `ufixed` value, the precision information is held by its ABI type.